pkg/controller/statefulset/stateful_set_utils.go

```diff
@@ -255,6 +281,124 @@ func newStatefulSetPod(set *apps.StatefulSet, ordinal int) *v1.Pod {
 	return pod
 }
 
+// newVersionedStatefulSetPod creates a new Pod for a StatefulSet. currentSet is the representation of the set at the
+// current revision. updateSet is the representation of the set at the updateRevision. currentRevision is the name of
+// the current revision. updateRevision is the name of the update revision. ordinal is the ordinal of the Pod. If the
+// returned error is nil, the returned Pod is valid.
+func newVersionedStatefulSetPod(currentSet *apps.StatefulSet, updateSet *apps.StatefulSet, currentRevision string, updateRevision string, ordinal int) *v1.Pod {
+	if (currentSet.Spec.UpdateStrategy.Type == apps.RollingUpdateStatefulSetStrategyType &&
+		ordinal < int(currentSet.Status.CurrentReplicas)) ||
+		(currentSet.Spec.UpdateStrategy.Type == apps.PartitionStatefulSetStrategyType &&
+			ordinal < int(currentSet.Spec.UpdateStrategy.Partition.Ordinal)) {
+		pod := newStatefulSetPod(currentSet, ordinal)
+		setPodRevision(pod, currentRevision)
+		return pod
+	}
+	pod := newStatefulSetPod(updateSet, ordinal)
+	setPodRevision(pod, updateRevision)
+	return pod
+}
+
+// getPatch returns a strategic merge patch that can be applied to restore a StatefulSet to a
+// previous version. If the returned error is nil the patch is valid. The current state that we save is just the
+// PodSpecTemplate. We can modify this later to encompass more state (or less) and remain compatible with previously
+// recorded patches.
```

Comment:
If a patch semantically means that you will roll back only the template at the time the revision is recorded it can still be applied to a statefulset where volume claims are also recorded. Rolling back will apply the prior revision without breaking versioning.